File: 11.py
class Solution:
    def maxArea(self, height: list[int]) -> int:
        max_area = 0
        p = 0
        q = len(height) - 1
        while p < q:
            area = min(height[p], height[q]) * (q - p)
            max_area = max(max_area, area)
            if height[p] < height[q]:
                p += 1
            else:
                q -= 1
        return max_area
    
if __name__ == '__main__':
    s = Solution()
    print(s.maxArea(list(map(int, input().split()))))

File: 11.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution{
public:
    int maxArea(vector<int>& height){
        int max_area = 0;
        int *p = &height[0], *q = &height[height.size()-1];
        while(p < q){
            int area = (q - p) * min(*p, *q);
            max_area = max_area > area ? max_area : area;
            if(*p < *q) p++;
            else q--;
        }
        return max_area;
    }
};

int main(){
    int n;
    cin >> n;
    vector<int> height(n);
    for(int i = 0; i < n; i++)
        cin >> height[i];
    Solution s;
    cout << s.maxArea(height) << endl;
    return 0;
}